pkg/controllers/node/termination/terminator/terminator.go

```diff
@@ -175,3 +207,145 @@ func (t *Terminator) podDeleteTimeWithGracePeriod(nodeGracePeriodExpirationTime
+		if ownerRef.Controller != nil && ownerRef.Kind == "ReplicaSet" {
+			rs := &appsv1.ReplicaSet{}
+			if err := t.kubeClient.Get(ctx, client.ObjectKey{Name: ownerRef.Name, Namespace: pod.Namespace}, rs); err != nil {
+				return nil, fmt.Errorf("get ReplicaSet: %w", err)
+			}
+			return rs, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func (t *Terminator) getOwnerDeployment(ctx context.Context, rs *appsv1.ReplicaSet) (*appsv1.Deployment, error) {
+	for _, ownerRef := range rs.GetOwnerReferences() {
+		if ownerRef.Controller != nil && ownerRef.Kind == "Deployment" {
+			deployment := &appsv1.Deployment{}
+			if err := t.kubeClient.Get(ctx, client.ObjectKey{Name: ownerRef.Name, Namespace: rs.Namespace}, deployment); err != nil {
+				return nil, fmt.Errorf("get Deployment: %w", err)
+			}
+			return deployment, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func (t *Terminator) RestartDeployments(ctx context.Context, deployments []*appsv1.Deployment, nodeName string) error {
+	var updateErrors []error
+
+	for _, deployment := range deployments {
+		if deployment.Spec.Template.Annotations == nil {
+			deployment.Spec.Template.Annotations = make(map[string]string)
+		}
+		restartedNode, exists := deployment.Spec.Template.Annotations[v1.DeploymentRestartNodeAnnotationKey]
+		if exists && restartedNode == nodeName {
+			continue
+		}
+
+		t.Lock()
+		if t.nodeRestartDeployments[nodeName] == nil {
+			t.nodeRestartDeployments[nodeName] = make(map[string]struct{})
+		}
+		t.nodeRestartDeployments[nodeName][deployment.Namespace+"/"+deployment.Name] = struct{}{}
+		t.Unlock()
+
+		deployment.Spec.Template.Annotations[v1.DeploymentRestartNodeAnnotationKey] = nodeName
+		if err := t.kubeClient.Update(ctx, deployment); err != nil {
+			updateErrors = append(updateErrors, err)
+			continue
+		}
+
+	}
+
+	if len(updateErrors) > 0 {
+		return fmt.Errorf("failed to restart some deployment: %v", updateErrors)
+	}
+
+	return nil
+}
+
+func (t *Terminator) GetRestartdeploymentsAndDrainPods(ctx context.Context, pods []*corev1.Pod, nodeName string) ([]*appsv1.Deployment, []*corev1.Pod, error) {
+	var drainPods []*corev1.Pod
+	var restartDeployments []*appsv1.Deployment
+	nodeDeploymentReplicas := make(map[string]int32)
+	deploymentCache := make(map[string]*appsv1.Deployment)
+	uniqueDeployments := make(map[string]struct{})
+
+	for _, pod := range pods {
+		deployment, err := t.getDeploymentFromCache(ctx, pod, deploymentCache)
+		if err != nil {
+			return nil, nil, err
+		}
+		if deployment != nil {
+			nodeDeploymentReplicas[deployment.Namespace+"/"+deployment.Name]++
+		}
+	}
+
+	for _, pod := range pods {
+		deployment := deploymentCache[pod.Namespace+"/"+pod.Name]
+
+		if deployment != nil {
+			key := deployment.Namespace + "/" + deployment.Name
+			if nodeDeploymentReplicas[key] >= *deployment.Spec.Replicas {
+				// If a deployment has multiple pods on this node, there will be multiple deployments here, and deduplication is required.
```

Comment:
If it's in the scaling-up process, this checking maybe wrong. how about check the ready replicas?